lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+					// Not enough data to fill the full buffer.
+					// Keep the readRequest for the next attempt.
+					
+					_this._DEBUG_LOG('TemplatedHTTPRequestData consume: wait',
+						'maxBytesToRead ===', maxBytesToRead,
+						'fillFullBuffer ===', fillFullBuffer,
+						'bytesRead ===', bytesRead,
+						'remainingBytesToRead ===', remainingBytesToRead,
+						'internalBuffersLength ===', internalBuffersLength
+					);
+					
+					return;
+				}
+				
+				// Fill the output buffer from the internal buffers.
+				while (internalBuffersLength > 0 && bytesRead < maxBytesToRead) {
+					buffer = internalBuffers[0];
+					var bytesReadThisTime = Math.min(remainingBytesToRead, buffer.length, internalBuffersLength);
+					buffer.copy(_this.buffer_, ptr, 0, bytesReadThisTime);
+					ptr += bytesReadThisTime;
+					bytesRead += bytesReadThisTime;
+					remainingBytesToRead = (maxBytesToRead - bytesRead);
+					if (bytesReadThisTime < buffer.length) {
+						internalBuffers[0] = buffer.slice(bytesReadThisTime);
+					}
+					else {
+						internalBuffers.shift();
+					}
+					internalBuffersLength -= bytesReadThisTime;
+				}
+			}
+			
+			if (bytesRead > 0 || internalBuffersClosed) {
+				// Have written to the buffer, the read should return the data length.
+				
+				_this._DEBUG_LOG('TemplatedHTTPRequestData consume: resolve', 
+					'maxBytesToRead ===', maxBytesToRead,
+					'fillFullBuffer ===', fillFullBuffer,
+					'bytesRead ===', bytesRead,
+					'remainingBytesToRead ===', remainingBytesToRead,
+					'internalBuffersLength ===', internalBuffersLength
+				);
+				
+				var deferred = readRequest.deferred;
+				readRequest = null;
+				deferred.resolve(bytesRead);
+			}
+			else {
+				// Keep the readRequest for the next attempt.
+				
+				_this._DEBUG_LOG('TemplatedHTTPRequestData consume: wait', 
+					'maxBytesToRead ===', maxBytesToRead,
+					'fillFullBuffer ===', fillFullBuffer,
+					'bytesRead ===', bytesRead,
+					'remainingBytesToRead ===', remainingBytesToRead,
+					'internalBuffersLength ===', internalBuffersLength
+				);
+				
+				return;
+			}
+		}
+		
+		
+		function onData(data) {
+			_this._DEBUG_LOG('TemplatedHTTPRequestData onData:', data.toString('utf8'));
+			
+			// Collect the data as it comes until it is explicitly read.
+			internalBuffers.push(data);
+			internalBuffersLength += data.length;
+			
+			consume();
+		}
+		
+		function onEnd() {
+			_this._DEBUG_LOG('TemplatedHTTPRequestData onEnd.');
+			
+			// Connection has closed, the next read should return zero bytes.
+			internalBuffersClosed = true;
+			
+			consume();
+		}
+		
+		function onError(err) {
+			_this._DEBUG_LOG('TemplatedHTTPRequestData onError:', err);
+			
+			// An error occurred, the next read should result in an exception.
+			internalError = err;
+			
+			consume();
+		}
+		
+		
+		// The algorithm is broken into async parts.
+		part1();
+		
+		function part1() {
```

Comment:
Nit: I'd call these functions `stepTemplatedParseHTTPResponseDataPart1()` or something. Since 1) `step...()` is a great prefix, and 2) it's great when function name is searchable in the code, while `part1` is not unique.